core/dbt/contracts/graph/manifest.py

```diff
@@ -169,7 +170,33 @@ def find(
     ):
         unique_id = self.get_unique_id(key, package, version)
         if unique_id is not None:
-            return self.perform_lookup(unique_id, manifest)
+            node = self.perform_lookup(unique_id, manifest)
+            if version is None and node.is_versioned and get_node_info():
+                # This was an 'unpinned' reference, resolved to latest_version.
+                # Check to see if newer versions are available, and log about it.
+                max_version: Union[str, float, None] = max(  # type: ignore
+                    [
+                        v.version
+                        for k, v in manifest.nodes.items()
+                        if k.startswith(f"{node.unique_id[:-len(str(node.version))]}")
+                    ]
+                )
```

Comment:
think this can be simplified to: 
```python
                max_version: Union[str, float, None] = max(  # type: ignore
                    [
                        v.version
                        for v in manifest.nodes.values()
                        if v.name == node.name
                    ]
                )
```

Reply:
this is ... clearly better

Reply:
Broader refactoring suggestion: using `UnparsedVersion` for the comparisons, similar to [what we do in the graph selector method here](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/graph/selector_methods.py#L661)

```
                max_version = max(
                    [
                        UnparsedVersion(v.version)
                        for v in manifest.nodes.values()
                        if v.name == node.name and v.version is not None
                    ]
                )
                assert node.latest_version
                if max_version > UnparsedVersion(node.latest_version):
                    fire_event(
                        UnpinnedRefNewVersionAvailable(
                            node_info=get_node_info(),
                            ref_node_name=node.name,
                            ref_node_package=node.package_name,
                            ref_node_version=str(node.version),
                            ref_max_version=str(max_version.v),
                        )
                    )
```

The TypeError issues in unit testing should be fixed by updating the MockNode constructor: 
```
+++ b/test/unit/utils.py
@@ -350,6 +350,7 @@ def MockNode(package, name, resource_type=None, **kwargs):
         **kwargs,
     )
     node.name = name
+    node.is_versioned = resource_type is NodeType.Model and version is not None
     return node
```

Reply:
yes!!!

Reply:
at the risk of expanding the intended scope of a partial parsing test - I'm going to borrow your commit 84cc3cabb69e704b0e09737121f3f08a90aabc24 and check that this message is being logged in the case of an unpinned `ref` to a latest-not-newest model